Allow nested columns & expanders.

ZeroCool940711 opened the pull request:
We should be able to use nested columns and expanders if we want to, it's up to the developer to decide if using them or not and it should not be disabled just because.

ZeroCool940711 committed:
Allow nested columns & expanders.

We should be able to use nested columns and expanders if we want to, it's up to the developer to decide if using them or not and it should not be disabled just because.

willhuang1997 commented:
Hi @ZeroCool940711 , Thanks for submitting a PR.

this was designed for a reason. If we allow that to happen, it would cause us to try to deal with more issues and more comments on discuss.streamlit.io and allow for behavior that we intend to not allow. I would recommend making an enhancement and trying to get people to comment / react as we do use those to prioritize our work.

ZeroCool940711 commented:
Hi @willhuang1997, sorry I wasn't able to reply sooner.

I understand your point but nested columns are a must if we want to have layouts with things horizontally and not just a bunch of vertical stuff on top of each other, columns are the best way to customize the layout to put things next to each other and also to save space, by not allowing nested columns there is a lot of stuff that we then will have to ways to do it differently. A library that you have to fight constantly and is not intuitive is not a good thing, and to give you an example, here is a simple app I was trying to do where I wanted to use nested columns:
![image](https://user-images.githubusercontent.com/5977640/187529387-87cc5b98-4463-453a-9202-05e523e1225b.png)

The idea with the app was to create an exact clone one that was made with Gradio, by having the exact same layout and design users would be able to easily switch from the gradio version to the streamlit version by just running the associated script in case they didn't like one of them or for any other reason, because I was not able to use nested columns I was not able to place the buttons one next to the other, I was able to put the buttons the way I wanted after modifying the code, but without that I had then to find another way to do that by using CSS but for me that was a waste of too much time so in the end I decided to just put the buttons one on top of each other, I end up changing the UI layout and making it different to the original, after I modified the code and removed the limitation placed in the code I was able to do something similar to what I originally wanted but at the end I gave up as I would then need to include an additional dependency unless you guys approve my PR. 

There were other cases where I wanted to use columns to organize things because it was easy for me but I was then limited by streamlit. One thing I liked when I first found and started using streamlit was that it was simple and worked like just writing python code and you didn't need anything extra like CSS or JS to do basic stuff unless you really wanted to, not been able to use columns inside another column or for that matter not been able to use any other element inside the same element just because it's not allowed and not because it won't work or because its impossible is just counterintuitive.

jrieke commented:
Hey @ZeroCool940711!

Thanks for your detailed feedback. As Will mentioned, we are restricting this by design today, mostly to prevent users from creating weird/too complex layouts. We don't want `st.columns` to become what HTML tables were 10-20 years ago, when they were abused for *any* kind of layout tweak.

Now, your use case is totally valid though and there have been a bunch of similar requests. We are aware of that and had some internal discussions recently if 1) we should allow columns inside columns or 2) prioritize another layout primitive (e.g. `st.grid`) to solve these use cases. I created an issue (#5284) to track your input here and get feedback from users. Please feel free to comment there! I'll also copy your screenshot above to our internal tools.

I know that it's frustrating that we are not "just merging". But we want to make Streamlit an amazing and simple tool, and so we are really careful not add things, where 3 months later we might think "hm that actually isn't the best way to solve this problem". (That's the down-side of a public open-source framework - once something is out there, you can't roll it back anymore without breaking a bunch of users!)

blackary commented:
For visibility: you can try out a monkey-patched version of this PR with this package: https://github.com/joy13975/streamlit-nested-layout
